pkg/controller/shared_state.go

```diff
@@ -1564,3 +1571,60 @@ func (c *SharedState) getNodesTopologySelector(ctx context.Context, nodes []v1.N
 	}
 	return topologySelector
 }
+
+func (s *SharedState) addNodeToAvailableAttachmentsMap(ctx context.Context, node string) {
+	nodeObj := &v1.Node{}
+	if err := s.cachedClient.Get(ctx, types.NamespacedName{Name: node}, nodeObj); err != nil {
+		capacity, _ := azureutils.GetNodeMaxDiskCountWithLabels(nodeObj.Labels)
+		var count atomic.Int32
+		count.Store(int32(capacity))
+		s.availableAttachmentsMap.Store(node, &count)
+		return
+	}
+	capacity, _ := azureutils.GetNodeRemainingDiskCount(ctx, s.cachedClient, node)
+	var count atomic.Int32
+	count.Store(int32(capacity))
+	s.availableAttachmentsMap.Store(node, &count)
+}
+
+func (s *SharedState) deleteNodeFromAvailableAttachmentsMap(ctx context.Context, node string) {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	w.Logger().V(5).Infof("Delete node (%s) from availableAttachmentsMap", node)
+	s.availableAttachmentsMap.Delete(node)
+}
+func (s *SharedState) decrementAttachmentCount(ctx context.Context, node string) bool {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	remainingCapacity, nodeExists := s.availableAttachmentsMap.Load(node)
+	if nodeExists && remainingCapacity != nil {
+		currentCapacity := int32(0)
+		for {
+			currentCapacity = remainingCapacity.(*atomic.Int32).Load()
+			if currentCapacity == int32(0) || remainingCapacity.(*atomic.Int32).CompareAndSwap(currentCapacity, currentCapacity-1) {
+				if currentCapacity == int32(0) {
+					err := status.Error(codes.OutOfRange, fmt.Sprintf("Current Capacity of disk attachment is 0. Can't decrement attachment count anymore."))
+					w.Logger().Errorf(err, "failed to decrement attachment count for node (%s) ", node)
+					return false
+				} else {
+					return true
+				}
+			}
+		}
+	}
+
+	err := status.Error(codes.FailedPrecondition, fmt.Sprintf("Node does not exist in availableAttachmentsMap."))
+	w.Logger().Errorf(err, "Can't decrement attachment count for the node (%s).", node)
```

Comment:
```go
err := errors.New("node not found")
w.Logger().Error(err, "Failed to decrement attachment count", consts.NodeNameLabel, node)
```

Reply:
Do you mean: 
w.Logger().Errorf(err, "Failed to decrement attachment count", consts.NodeNameLabel, node)

I've checked the implementation of Errorf() and Error(). It seems Error() can only have 2 arguments. Below is the original code:
func (l Logger) Errorf(err error, msg string, params ...interface{}) {
	l.WithCallDepth(1).Error(err, fmt.Sprintf(msg, params...))
}

Reply:
Ah, my bad. I forgot that the workflow logger interface does not implement the additional `keysAndValues` parameters that the [`logr.Logger.Error`](https://pkg.go.dev/github.com/go-logr/logr#Logger.Error) method does.